stdlib/LinearAlgebra/src/matmul.jl

```diff
@@ -454,8 +449,11 @@ end
     A
 end
 
+promote_unless_bool(α::Number, ::Type{T}) where {T<:Number} = first(promote(α, zero(T)))
+promote_unless_bool(α::Bool, ::Type{T}) where {T<:Number} = α
```

Comment:
I'm just curious. Why this is necessary to handle a `Bool` separately? I thought it would be unnecessary since the promotion would happen via `ccall` anyway.

Reply:
I don't know. I just saw that the `BLAS.gemm!` and alike accept both `T` and `Bool`, so I thought they'd be able to handle `Bool`s without conversion. I want to avoid that we allocate new variables for the promoted versions. Also, I had the impression that most of the test failures of my first commit was related to `herk!`, and the fact that one the coeffs was `Complex`. I assume I had promoted the coefficient to the complex eltype of the matrices, but `herk!` doesn't accept that? Now that we have much less repetition, I could go back and try.

Reply:
I extended `gemm!` etc. to accept `Bool` alpha and beta just for making 5-arg `mul!` easier. There was nothing fancy trick. `gemm!` etc. special case `beta = 0` internally.

Reply:
So BLAS promotes the `Bool` values to numeric values internally? Anyway, I got rid of the entire `promote_unless_bool` thing. I hope BLAS does check internally for `isone`, and in that case does not multiply with 1, nor adds 0.

Reply:
I think BLAS (Fortran code) checks `beta = 0`. At least that's the behavior I see with `BLAS.gemm!('N', 'N', 1.0, ones(10, 10), ones(10, 10), 0.0, fill(NaN, 10, 10))`.